### PR TITLE
[ARCTIC-1336]: Support mixed iceberg in external catalog based on iceberg catalog. 

### LINKED ISSUE #1336: [Subtask]: Support creating mixed-iceberg format tables in any catalog which support iceberg.
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/arctic/issues?q=is%3Aissue) and found no similar issues.

### Description

- Support management iceberg and mixed-iceberg table in the external catalog

### Parent issue

#1061 

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed?

fix #1336 

This PR refactors the MixedIceberg format tables, making mixed-iceberg table implements based on Iceberg Catalog instead of HadoopTables. So that we can create a mixed-iceberg table on any catalog type if it could create an iceberg table.

## Brief change log

- Add a new ArcticCatalog implement `com.netease.arctic.mixed.catalog.BasicMixedIcebergCatalog`  
- Support mixed-iceberg format in all metastore types ( Hive, Hadoop, Custom)  with a `BasicMixedIcebergCatalog` implement.
- The `PrimaryKeySpec` is stored in table properties of the base store(Iceberg table), not the AMS system db.

## How was this patch tested?
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] Run test locally before making a pull request

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/catalog/MixedIcebergCatalogImpl.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.netease.arctic.ams.api.CatalogMeta;
+import com.netease.arctic.ams.api.TableIdentifier;
+import com.netease.arctic.mixed.catalog.BasicMixedIcebergCatalog;
+import com.netease.arctic.table.ArcticTable;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class MixedIcebergCatalogImpl extends ExternalCatalog {
+  BasicMixedIcebergCatalog mixedIcebergCatalog;
```
Comment (zhoujinsong) [inline ams/server/src/main/java/com/netease/arctic/server/catalog/MixedIcebergCatalogImpl.java:29]: Would it be more appropriate to define it as private here?

File: core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java [hunk 2]
```diff
@@ -114,22 +104,23 @@ public static ArcticCatalog load(AmsClient client, String catalogName, Map<Strin
...
-            catalogImpl = AMS_CATALOG_IMPL;
+            catalogImpl = MIXED_ICEBERG_CATALOG_IMP;
           }
           break;
         case CATALOG_TYPE_HIVE:
           if (tableFormat.equals(TableFormat.ICEBERG)) {
             catalogImpl = ICEBERG_CATALOG_IMPL;
           } else if (tableFormat.equals(TableFormat.MIXED_HIVE)) {
             catalogImpl = HIVE_CATALOG_IMPL;
+          } else if (TableFormat.MIXED_ICEBERG == tableFormat) {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java:120]: Although using both `==` and `equals` would work here, I prefer to keep the approach consistent.

File: core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java [hunk 3]
```diff
@@ -166,6 +161,20 @@ public static ArcticCatalog load(AmsClient client, String catalogName, Map<Strin
     }
   }
 
+
+  public static CatalogMeta loadMeta(String catalogUrl) {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/catalog/CatalogLoader.java:165]: Can you add some java doc for this method.

File: core/src/main/java/com/netease/arctic/catalog/MixedTables.java [hunk 4]
```diff
@@ -137,17 +136,19 @@ protected UnkeyedTable loadUnKeyedTable(TableMeta tableMeta) {
         fileIO, tableMetaStore.getConfiguration()), fileIO, amsClient, catalogMeta.getCatalogProperties());
   }
 
-  public ArcticTable createTableByMeta(TableMeta tableMeta, Schema schema, PrimaryKeySpec primaryKeySpec,
-                                       PartitionSpec partitionSpec) {
+  public ArcticTable createTableByMeta(
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/catalog/MixedTables.java:139]: Can you revert these changes to the code format? The old codes seemed good here.

File: core/src/main/java/com/netease/arctic/io/ArcticFileIOIcebergAdapter.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.SupportsPrefixOperations;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+public class ArcticFileIOIcebergAdapter implements ArcticFileIO {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/io/ArcticFileIOIcebergAdapter.java:12]: What's the difference of this class and `ArcticFileIOAdapter`?

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 6]
```diff
@@ -257,6 +256,20 @@ private TableProperties() {
 
   public static final String OWNER = "owner";
 
+  /**
+   * table format related properties
+   */
+  public static final String TABLE_FORMAT = "table-format";
+  public static final String TABLE_FORMAT_MIXED_ICEBERG = "mixed-iceberg";
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/table/TableProperties.java:263]: Can we resue the enum `com.netease.arctic.ams.api.TableFormat` rather than defining too many property values?

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 7]
```diff
@@ -257,6 +256,20 @@ private TableProperties() {
...
+  public static final String TABLE_FORMAT_MIXED_HIVE = "mixed-hive";
+
+  public static final String MIXED_FORMAT_PRIMARY_KEY_FIELDS = "mixed-format.primary-key-fields";
+  public static final String MIXED_FORMAT_TABLE_STORE = "mixed-format.table-store";
+  public static final String MIXED_FORMAT_TABLE_STORE_BASE = "mixed-format.table-store.base";
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/table/TableProperties.java:268]: Would it be more appropriate to use shorter names for property values, such as `base` and `change`?

File: ams/server/src/main/java/com/netease/arctic/server/catalog/CatalogBuilder.java [hunk 8]
```diff
@@ -25,21 +22,24 @@ public static ServerCatalog buildServerCatalog(CatalogMeta catalogMeta, Configur
...
           return new IcebergCatalogImpl(catalogMeta);
         } else if (tableFormat.equals(TableFormat.MIXED_HIVE)) {
           return new MixedHiveCatalogImpl(catalogMeta);
+        } else if (TableFormat.MIXED_ICEBERG == tableFormat) {
+          return new MixedIcebergCatalogImpl(catalogMeta);
         } else {
-          throw new IllegalArgumentException("Hive Catalog support iceberg table and mixed hive table only");
+          throw new IllegalArgumentException("Hive Catalog support iceberg/mixed-hive/mixed-iceberg table only");
         }
       case CATALOG_TYPE_AMS:
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/catalog/CatalogBuilder.java:44]: It's auto-formatted by IDEA.

**Review by zhoujinsong — APPROVED** [review]
LGTM.
